File: examples/full-flow.ts
/**
 * @cinchor/sdk — end-to-end example + smoke test.
 *
 * Proves the SDK against a live Omne node by driving the full lifecycle:
 *   mint capability → enforce (in-bounds, allowed) → enforce (over-budget, refused)
 *   → attest decision → verify (tamper-evident) → audit read → revoke → enforce (refused)
 *
 * It is both a runnable example and a smoke test: every step asserts the expected
 * outcome and the process exits non-zero on any mismatch.
 *
 * Run (against a local node with the contract deployed + wallets funded):
 *   CINCHOR_CONTRACT_ADDRESS=om1z… \
 *   CINCHOR_WALLETS=/path/to/wallets.json \
 *   npm run smoke
 *
 * Env:
 *   CINCHOR_RPC_URL          default http://127.0.0.1:26657
 *   CINCHOR_CHAIN_ID         default 3 (Ignis)
 *   CINCHOR_CONTRACT_NAME    default cinchor_permissions
 *   CINCHOR_CONTRACT_ADDRESS required — the deployed contract's om1z address
 *   CINCHOR_WALLETS          required — JSON: { principal:{mnemonic}, agent:{mnemonic} }
 */

import { readFileSync } from 'node:fs';
import { Wallet } from '@omne/sdk';
import {
  CinchorClient,
  EnforcementCode,
  CapabilityStatus,
  Verdict,
  nowSecs,
  type Signer,
} from '../dist/index.js';

// ── Tiny assertion harness ──────────────────────────────────────────
let passed = 0;
let failed = 0;
function check(label: string, cond: boolean, detail = ''): void {
  if (cond) {
    passed++;
    console.log(`  ✅ ${label}${detail ? `  (${detail})` : ''}`);
  } else {
    failed++;
    console.log(`  ❌ ${label}${detail ? `  (${detail})` : ''}`);
  }
}

// ── Config ──────────────────────────────────────────────────────────
function requireEnv(name: string): string {
  const v = process.env[name];
  if (!v) throw new Error(`${name} is required`);
  return v;
}

const RPC_URL = process.env.CINCHOR_RPC_URL ?? 'http://127.0.0.1:26657';
const CHAIN_ID = Number(process.env.CINCHOR_CHAIN_ID ?? 3);
const CONTRACT_NAME = process.env.CINCHOR_CONTRACT_NAME ?? 'cinchor_permissions';
const CONTRACT_ADDRESS = requireEnv('CINCHOR_CONTRACT_ADDRESS');
const WALLETS_PATH = requireEnv('CINCHOR_WALLETS');

function signerFromMnemonic(mnemonic: string): Signer {
  return Wallet.fromMnemonic(mnemonic).getAccount(0) as unknown as Signer;
}

async function main(): Promise<void> {
  const wallets = JSON.parse(readFileSync(WALLETS_PATH, 'utf8')) as {
    principal: { mnemonic: string };
    agent: { mnemonic: string };
  };
  const principal = signerFromMnemonic(wallets.principal.mnemonic);
  const agent = signerFromMnemonic(wallets.agent.mnemonic);

  console.log(`\nCinchor end-to-end smoke`);
  console.log(`  rpc:       ${RPC_URL} (chain ${CHAIN_ID})`);
  console.log(`  contract:  ${CONTRACT_NAME} @ ${CONTRACT_ADDRESS}`);
  console.log(`  principal: ${principal.address}`);
  console.log(`  agent:     ${agent.address}\n`);

  const cinchor = await CinchorClient.connect({
    network: { name: 'ignis', chainId: CHAIN_ID, rpcUrl: RPC_URL },
    contract: { name: CONTRACT_NAME, address: CONTRACT_ADDRESS },
  });

  // 1. Mint a scoped capability: ceiling 100, valid 1h, no allowlist.
  console.log('1. mintCapability (principal grants agent a capability)');
  const { capabilityId } = await cinchor.mintCapability({
    principal,
    agent: agent.address,
    maxSpend: 100n,
    ttlSeconds: 3600,
  });
  const minted = await cinchor.getCapability(capabilityId);
  check('capability is active after mint', minted.status === CapabilityStatus.Active, minted.statusLabel);
  check('maxSpend recorded', minted.maxSpend === 100n, `maxSpend=${minted.maxSpend}`);
  console.log(`     capabilityId: ${capabilityId}`);

  // 2. In-bounds action → allowed.
  console.log('2. enforce 40 (in bounds)');
  const a1 = await cinchor.enforce({ capability: capabilityId, agent, amount: 40n });
  check('first in-bounds action allowed', a1.allowed, a1.reason);

  // 3. Second in-bounds action → allowed (total 80/100).
  console.log('3. enforce 40 (in bounds, total 80/100)');
  const a2 = await cinchor.enforce({ capability: capabilityId, agent, amount: 40n });
  check('second in-bounds action allowed', a2.allowed, a2.reason);

  // 4. Over-budget action → refused by the substrate (80 + 50 > 100).
  console.log('4. enforce 50 (over budget → must be refused)');
  const a3 = await cinchor.enforce({ capability: capabilityId, agent, amount: 50n });
  check('over-budget action refused', !a3.allowed, a3.reason);
  check('refusal reason is over_budget', a3.code === EnforcementCode.OverBudget, a3.reason);

  // 5. State reflects exactly the two committed actions.
  const afterActions = await cinchor.getCapability(capabilityId);
  check('totalSpent = 80 (refusal did not mutate)', afterActions.totalSpent === 80n, `totalSpent=${afterActions.totalSpent}`);
  check('actionCount = 2', afterActions.actionCount === 2n, `actionCount=${afterActions.actionCount}`);

  // 6. Attest a decision; verify it is tamper-evident.
  console.log('5. attest a decision + verify (tamper-evidence)');
  const decision = { model: 'demo-triage-v1', input: { claim: 'A-123' }, output: 'approve', at: Number(nowSecs()) };
  const { attestationId, contextHash } = await cinchor.attest({
    capability: capabilityId,
    agent,
    context: decision,
    verdict: Verdict.InPolicy,
  });
  const good = await cinchor.verifyAttestation(decision, attestationId);
  check('attestation verifies against the original context', good.ok, `hash=${contextHash.slice(0, 14)}…`);
  const tampered = await cinchor.verifyAttestation({ ...decision, output: 'deny' }, attestationId);
  check('tampered context fails verification', !tampered.ok, 'hash mismatch detected');

  // 7. Revoke, then prove the substrate refuses post-revocation actions.
  console.log('6. revoke + enforce (must be refused: revoked)');
  await cinchor.revoke({ capability: capabilityId, principal });
  const revokedState = await cinchor.getCapability(capabilityId);
  check('capability is revoked', revokedState.status === CapabilityStatus.Revoked, revokedState.statusLabel);
  const a4 = await cinchor.enforce({ capability: capabilityId, agent, amount: 1n });
  check('post-revocation action refused', !a4.allowed, a4.reason);
  check('refusal reason is revoked', a4.code === EnforcementCode.Revoked, a4.reason);

  // ── Summary ────────────────────────────────────────────────────
  console.log(`\n${failed === 0 ? '✅ PASS' : '❌ FAIL'} — ${passed} checks passed, ${failed} failed\n`);
  if (failed > 0) process.exit(1);
}

main().catch((err) => {
  console.error('\n❌ smoke run errored:', err);
  process.exit(1);
});

File: py/examples/full_flow.py
"""Cinchor Python SDK — end-to-end example + smoke test.

Drives the full lifecycle against a live Omne node + deployed contract:
  mint capability → enforce (in-bounds ×2) → enforce (over-budget, refused)
  → attest decision → verify (tamper-evident) → revoke → enforce (refused).
Every step asserts; the process exits non-zero on any mismatch.

  python full_flow.py <rpc_url> <contract_address> <wallets_json>
"""

import json
import sys

from omne_sdk import Wallet

from cinchor import (
    CapabilityStatus,
    CinchorClient,
    CinchorConfig,
    ContractConfig,
    EnforcementCode,
    NetworkConfig,
    Verdict,
    now_secs,
)

RPC, CONTRACT, WALLETS = sys.argv[1], sys.argv[2], sys.argv[3]
CHAIN_ID = 3
CONTRACT_NAME = "cinchor_permissions"

passed = failed = 0


def check(label, cond, detail=""):
    global passed, failed
    mark = "PASS" if cond else "FAIL"
    if cond:
        passed += 1
    else:
        failed += 1
    print(f"  [{mark}] {label}  ({detail})")


def main() -> int:
    w = json.load(open(WALLETS))
    principal = Wallet.from_mnemonic(w["principal"]["mnemonic"]).get_account(0)
    agent = Wallet.from_mnemonic(w["agent"]["mnemonic"]).get_account(0)
    cinchor = CinchorClient.connect(CinchorConfig(
        network=NetworkConfig("ignis", CHAIN_ID, RPC),
        contract=ContractConfig(name=CONTRACT_NAME, address=CONTRACT),
    ))
    print(f"\nCinchor Python smoke\n  contract: {CONTRACT_NAME} @ {CONTRACT}")
    print(f"  principal: {principal.address}\n  agent:     {agent.address}\n")

    print("1. mintCapability")
    minted = cinchor.mint_capability(principal=principal, agent=agent.address, max_spend=100, ttl_seconds=3600)
    cap = minted["capability_id"]
    state = cinchor.get_capability(cap)
    check("capability active after mint", state.status == CapabilityStatus.ACTIVE, state.status_label)
    check("maxSpend recorded", state.max_spend == 100, f"maxSpend={state.max_spend}")
    print(f"     capabilityId: {cap}")

    print("2. enforce 40 (in bounds)")
    a1 = cinchor.enforce(capability=cap, agent=agent, amount=40)
    check("first in-bounds action allowed", a1.allowed, a1.reason)

    print("3. enforce 40 (in bounds, total 80/100)")
    a2 = cinchor.enforce(capability=cap, agent=agent, amount=40)
    check("second in-bounds action allowed", a2.allowed, a2.reason)

    print("4. enforce 50 (over budget → refused)")
    a3 = cinchor.enforce(capability=cap, agent=agent, amount=50)
    check("over-budget action refused", not a3.allowed, a3.reason)
    check("refusal reason is over_budget", a3.code == EnforcementCode.OVER_BUDGET, a3.reason)
    after = cinchor.get_capability(cap)
    check("totalSpent = 80 (refusal did not mutate)", after.total_spent == 80, f"totalSpent={after.total_spent}")
    check("actionCount = 2", after.action_count == 2, f"actionCount={after.action_count}")

    print("5. attest a decision + verify (tamper-evidence)")
    decision = {"model": "demo-triage-v1", "input": {"claim": "A-123"}, "output": "approve", "at": now_secs()}
    att = cinchor.attest(capability=cap, agent=agent, context=decision, verdict=Verdict.IN_POLICY)
    good = cinchor.verify_attestation(decision, att["attestation_id"])
    check("attestation verifies against the original context", good["ok"], f"hash={att['context_hash'][:14]}…")
    tampered = cinchor.verify_attestation({**decision, "output": "deny"}, att["attestation_id"])
    check("tampered context fails verification", not tampered["ok"], "hash mismatch detected")

    print("6. revoke + enforce (refused: revoked)")
    cinchor.revoke(capability=cap, principal=principal)
    revoked = cinchor.get_capability(cap)
    check("capability revoked", revoked.status == CapabilityStatus.REVOKED, revoked.status_label)
    a4 = cinchor.enforce(capability=cap, agent=agent, amount=10)
    check("post-revocation action refused", not a4.allowed, a4.reason)
    check("refusal reason is revoked", a4.code == EnforcementCode.REVOKED, a4.reason)

    print(f"\n{'PASS' if failed == 0 else 'FAIL'} — {passed} checks passed, {failed} failed\n")
    return 0 if failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
